### PR TITLE
Add haskell language layer

### PR DESCRIPTION
**Pull request by DanTheColoradan**
Status: ready for review.

~~Todo: I need feedback and to use all the features to make sure they work.~~

### CODE REVIEW

**Comment by DanTheColoradan** [comment]
@geksilla thanks.

I'm considering on changing keybindings to be:

```md
<<<<<<< Updated upstream
| <kbd>SPC m h t</kbd>        | show type                         |
| <kbd>SPC m h i</kbd>        | show info                         |
| <kbd>SPC m h l</kbd>        | hoogle lookup                     |
| <kbd>SPC m j d</kbd>        | go to declaration                 |
| <kbd>SPC m i t</kbd>        | insert type                       |
| <kbd>SPC m i i</kbd>        | insert import                     |
| <kbd>SPC m c s</kbd>        | select target                     |
| <kbd>SPC m c b</kbd>        | build target                      |
| <kbd>SPC m c c</kbd>        | clean target                      |
| <kbd>SPC m c t</kbd>        | run target's tests                |
=======
| <kbd>SPC h t</kbd>          | show type                         |
| <kbd>SPC h i</kbd>          | show info                         |
| <kbd>SPC h l</kbd>          | hoogle lookup                     |
| <kbd>SPC j d</kbd>          | jump to declaration               |
| <kbd>SPC i t</kbd>          | insert type                       |
| <kbd>SPC i i</kbd>          | insert import                     |
| <kbd>SPC c s</kbd>          | select target                     |
| <kbd>SPC c b</kbd>          | build target                      |
| <kbd>SPC c c</kbd>          | clean target                      |
| <kbd>SPC c t</kbd>          | run target's tests                |
>>>>>>> Stashed changes
```

Part of this would depend on #144 changing back to *not* using `i`, `j`, or `h`. Specifically the conflicts are:
  - [i] -> +insert
  - [j] -> +jump
  - [h] -> +help

The part that doesn't depend on #144 is changing to `SPC c` instead of `SPC m c` (related to convo in #162).

Thoughts? /cc @jessejanderson @dvcrn

File: src/cljs/proton/layers/lang/haskell/core.cljs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+(defmethod get-packages :lang/haskell []
+ [:autocomplete-haskell
+   :haskell-ghc-mod
+   :haskell-hoogle
+   :ide-haskell
+   :ide-haskell-cabal
+   :language-haskell])
+
+(defmethod describe-mode :lang/haskell []
```
Comment (DanTheColoradan) [inline src/cljs/proton/layers/lang/haskell/core.cljs]: Need help here. Do I need to add cabal as a grammar and then add the `.cabal` extension? Auto-indent on cabal isn't working as I would expect and I'm not if it's my fault or downward dependency.

File: src/cljs/proton/layers/lang/haskell/core.cljs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+(defn- haskell-mode-init []
+ (atom-env/set-grammar "Haskell"))
+
+(defmethod init-layer! :lang/haskell
+  [_ {:keys [config layers]}]
+  (helpers/console! "init" :lang/haskell)
+  (register-layer-dependencies :tools/linter [:linter-hlint]))
+
+(defmethod get-keybindings :lang/haskell []
+  {:f {:category "files"
```
Comment (geksilla) [inline src/cljs/proton/layers/lang/haskell/core.cljs]: These keybindings should move to mode and not affect global keymap.

File: src/cljs/proton/layers/lang/haskell/core.cljs (hunk 1 above)
Comment (geksilla) [inline src/cljs/proton/layers/lang/haskell/core.cljs]: If you want to provide special key maps for `.cabal` files you need to define another mode using `proton.lib.mode/define-mode` function in `init-layer` hook.

 I don't know why auto-indent doesn't work. Need to check this PR.

**Comment by DanTheColoradan** [comment]
Thanks @geksilla. I have updated my branch and I'm going to test a bit longer before claiming it's merge ready.

**Comment by dvcrn** [comment]
:+1:

**Comment by geksilla** [comment]
nice work! :+1:

**Comment by DanTheColoradan** [comment]
Almost there. I thought I had figured it out, but I don't know how to get rid of the info gutter/pane ide-haskell provides. I still see a bar across the bottom, which is the top-bar when it is open:

screen shot 2016-03-03 at 12 06 17 pm

It's the light gray bar in the picture.

Any way to hide this?

**Comment by garrettm** [comment]
@DanTheColoradan I think spacemacs' hoogle lookup is `SPC m h h`, maybe use that here as well?  (I could be mis-remembering here..)

Are you testing it with Stack or Cabal?

**Comment by DanTheColoradan** [comment]
Hm. I haven't been consistent enough to claim I just used one. Is the bar not showing up when you test it? What are you thinking :) ?

**Comment by garrettm** [comment]
@DanTheColoradan I've not tested this layer yet (although I will when it's available!).

Before proton, I remember having similar issues (around the bottom bar) with atom haskell packages, though...

**Comment by dvcrn** [comment]
If the bar is a serious problem and something that is not needed at all, we could add a css rule to just hide it. Just a thought

**Comment by DanTheColoradan** [comment]
Thanks. Your comments spurred me to try again. It is a panel from `autocomplete-haskell` (autocomplete-haskell:toggle-completion-hint).

I'm going to fix it, push my changes, and then solicit final recommendations.

**Comment by DanTheColoradan** [comment]
Ok :dancer: it feels ready for another look over and then merging. Yay! Let me know what, if anything, anyone would like fixed / changed.

:beers:

**Comment by dvcrn** [comment]
LGTM! @geksilla any feedback? Otherwise I'd merge it

**Comment by geksilla** [comment]
:+1: good job!

**Pull request merged**